File: Algorithm/AlgorithmCSharp/AlgorithmCSharp/Q4.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace AlgorithmCSharp
{
    /// <summary>
    /// アナグラム発見
    /// </summary>
    class Q4
    {
        public void Start(string filePath)
        {
            var wordList = File.ReadLines(filePath);
            var anagramGroupList = this.ComputeAnagramGroupList(wordList);

            foreach (var group in anagramGroupList)
            {
                Console.WriteLine(group.Key + "：" + "{" + string.Join(", ", group) + "}");
            }
        }

        /// <summary>
        /// 単語の一覧をアナグラム別にまとめます。
        /// </summary>
        /// <param name="wordList"></param>
        /// <returns></returns>
        private IGrouping<string, string>[] ComputeAnagramGroupList(IEnumerable<string> wordList)
        {
            return wordList
                .GroupBy(
                    word => new string(word.OrderBy(c => c).ToArray()),
                    word => word
                )
                .ToArray();
        }
    }
}

File: Algorithm/AlgorithmFSharp/AlgorithmFSharp/Q4.fs
// アナグラム
module Q4

open System.IO

// 単語の一覧をアナグラム別にまとめます。
let computeAnagramGroupList (wordList: string seq) = 
    wordList |> Seq.groupBy(fun word ->  word.ToCharArray() |> Array.sort |> (fun c -> System.String c))

let start filePath =
    let wordList = File.ReadLines filePath
    let anagramGroupList = computeAnagramGroupList wordList

    anagramGroupList
    |> Seq.map(fun anagramGroup -> printfn "%A : %A" (fst anagramGroup) ((snd anagramGroup) |> (fun word -> System.String.Join(", ", word))))
    |> Seq.toList